node/authorship/src/authorship.rs

```diff
@@ -520,25 +548,87 @@ where
+        let client = self.client.clone();
+        let backend = self.backend.clone();
+        let pseudo_inherent = block_builder.create_terminal_extrinsic(self.max_gas)?;
+        let (extrinsics, api, version, _, _, estimated_header_size) =
+            block_builder.clone().deconstruct();
+        // We need the overlay changes and transaction storage cache to send to a new thread.
+        // The cloned `RuntimaApi` object can't be sent to a new thread directly so we have to
+        // break it down into parts (that are `Send`) and then reconstruct it in the new thread.
+        // If changes applied successfully, the updated extrinsics and api parts will be sent back
+        // to update the original block builder and finalize the block.
+        let (_, api_params) = api.deref().clone().deconstruct();
+
+        let update_block = async move {
+            let (tx, rx) = oneshot::channel();
+            let spawn_handle = self.spawn_handle.clone();
+
+            spawn_handle.spawn_blocking(
+                "block-builder-push",
+                None,
+                Box::pin(async move {
+                    debug!(target: "gear::authorship", "⚙️  Pushing Gear::run extrinsic into the block...");
+                    let mut block_builder = BlockBuilder::<'_, Block, C, B>::from_parts(
+                        extrinsics,
+                        ApiRef::from(C::Api::restore(client.as_ref(), api_params)),
+                        version,
+                        parent_hash,
+                        backend.as_ref(),
+                        estimated_header_size);
+                    let outcome = block_builder.push(pseudo_inherent).map(|_| {
+                        let (extrinsics, api, _, _, _, _) =
+                            block_builder.deconstruct();
+                        let (_, api_params) = api.deref().clone().deconstruct();
+                        (extrinsics, api_params)
+                    });
+                    if tx.send(outcome).is_err() {
+                        debug!(
+                            target: "gear::authorship",
+                            "🔒 Send failure: the receiver must have already closed the channel.");
+                    };
+                }),
+            );
 
-        debug!(target: "gear::authorship", "⚙️  Pushing Gear::run extrinsic into the block...");
-        match block_builder.push(custom_extrinsic) {
-            Ok(()) => {
-                debug!(target: "gear::authorship", "⚙️  ... pushed to the block");
-            }
-            Err(ApplyExtrinsicFailed(Validity(e))) if e.exhausted_resources() => {
-                warn!(target: "gear::authorship", "⚠️  Dropping terminal extrinsic from an overweight block.")
+            rx.await?
+        }.boxed();
+        match futures::future::select(
+            update_block,
+            // TODO: consider adding some tolerance here like `deadline` + 20% or something.
```

Comment:
We can add 20%, but keep in mind that spawning a thread consumes resources. Perhaps there is a way to benchmark   proposer.

Reply:
That's why I was thinking about relaxing the deadline. Not only it takes time to spawn a thread (which effectively reduces the time budget for the extrinsics processing) but also we have about 1/3 of the block time for finalization, and we know we don't have any heavy `on_finalize()` hooks in any of the pallets so letting the pseudo-inherent to run a little longer sounds reasonable.
But I agree that we can only tell how much longer exactly we can let it run for only after benchmarking.